var/spack/repos/builtin/packages/py-line-profiler/package.py

```diff
@@ -12,6 +12,9 @@ class PyLineProfiler(PythonPackage):
     homepage = "https://github.com/rkern/line_profiler"
     url      = "https://pypi.io/packages/source/l/line_profiler/line_profiler-2.0.tar.gz"
 
+    # unfortunately no new release is available and 2.0 does not
+    # yet work with python?3.5?: so we use a current commit
+    version('2.new', git='git@github.com:rkern/line_profiler.git', commit='3cdc2fab6793802d31130b1af2705bddebf6f7fd')
```

Comment:
Uh, 2.1.2 is available on pypi: https://pypi.org/project/line_profiler/#history

Python 3.5 support was added in 2.1.

Reply:
but not py37 right? Let me test

Reply:
Yes confirmed
```
==> Installing py-line-profiler
==> Searching for binary cache of py-line-profiler
==> Warning: No Spack mirrors are currently configured
==> No binary for py-line-profiler found: installing from source
==> Fetching https://pypi.io/packages/source/l/line_profiler/line_profiler-2.1.2.tar.gz
######################################################################## 100.0%
==> Staging archive: /var/folders/1q/jxx1cgcx3l13d153s6_c1zc00000gn/T/healther/spack-stage/py-line-profiler-2.1.2-btm3uhjlak4kxdc3qvujuw3zv3vciozc/line_profiler-2.1.2.tar.gz
==> Created stage in /var/folders/1q/jxx1cgcx3l13d153s6_c1zc00000gn/T/healther/spack-stage/py-line-profiler-2.1.2-btm3uhjlak4kxdc3qvujuw3zv3vciozc
==> No patches needed for py-line-profiler
==> Building py-line-profiler [PythonPackage]
==> Executing phase: 'build'
==> Error: ProcessError: Command exited with status 1:
    '/Users/healther/Documents/git.nosync/spack/opt/spack/darwin-mojave-broadwell/clang-10.0.0-apple/python-3.7.4-sbzjkpqd5iukn357yc53xkgqkrybgfdh/bin/python3.7' '-s' 'setup.py' '--no-user-cfg' 'build'

16 errors found in build log:
     9      copying line_profiler_py35.py -> build/lib.macosx-10.14.6-x86_64-3.7
     10     running build_ext
     11     skipping '_line_profiler.c' Cython extension (up-to-date)
     12     building '_line_profiler' extension
     13     creating build/temp.macosx-10.14.6-x86_64-3.7
     14     /Users/healther/Documents/git.nosync/spack/lib/spack/env/clang/clang -Wno-unused-result -Wsign-compare -Wunreachable-code -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -I/Users/healther/Documents/git.n
            osync/spack/opt/spack/darwin-mojave-broadwell/clang-10.0.0-apple/python-3.7.4-sbzjkpqd5iukn357yc53xkgqkrybgfdh/include/python3.7m -c _line_profiler.c -o build/temp.macosx-10.14.6-x86_64-3.7/_
            line_profiler.o
  >> 15     _line_profiler.c:7890:21: error: no member named 'exc_type' in 'struct _ts'
     16         *type = tstate->exc_type;
     17                 ~~~~~~  ^
  >> 18     _line_profiler.c:7891:22: error: no member named 'exc_value' in 'struct _ts'; did you mean 'curexc_value'?
     19         *value = tstate->exc_value;
     20                          ^~~~~~~~~
     21                          curexc_value
     22     /Users/healther/Documents/git.nosync/spack/opt/spack/darwin-mojave-broadwell/clang-10.0.0-apple/python-3.7.4-sbzjkpqd5iukn357yc53xkgqkrybgfdh/include/python3.7m/pystate.h:240:15: note: 'curex
            c_value' declared here
     23         PyObject *curexc_value;
     24                   ^
  >> 25     _line_profiler.c:7892:19: error: no member named 'exc_traceback' in 'struct _ts'; did you mean 'curexc_traceback'?
     26         *tb = tstate->exc_traceback;
     27                       ^~~~~~~~~~~~~
     28                       curexc_traceback
     29     /Users/healther/Documents/git.nosync/spack/opt/spack/darwin-mojave-broadwell/clang-10.0.0-apple/python-3.7.4-sbzjkpqd5iukn357yc53xkgqkrybgfdh/include/python3.7m/pystate.h:241:15: note: 'curex
            c_traceback' declared here
     30         PyObject *curexc_traceback;
     31                   ^
  >> 32     _line_profiler.c:7899:24: error: no member named 'exc_type' in 'struct _ts'
     33         tmp_type = tstate->exc_type;
     34                    ~~~~~~  ^
  >> 35     _line_profiler.c:7900:25: error: no member named 'exc_value' in 'struct _ts'; did you mean 'curexc_value'?
     36         tmp_value = tstate->exc_value;
     37                             ^~~~~~~~~
     38                             curexc_value
     39     /Users/healther/Documents/git.nosync/spack/opt/spack/darwin-mojave-broadwell/clang-10.0.0-apple/python-3.7.4-sbzjkpqd5iukn357yc53xkgqkrybgfdh/include/python3.7m/pystate.h:240:15: note: 'curex
            c_value' declared here
     40         PyObject *curexc_value;
     41                   ^
  >> 42     _line_profiler.c:7901:22: error: no member named 'exc_traceback' in 'struct _ts'; did you mean 'curexc_traceback'?
     43         tmp_tb = tstate->exc_traceback;
     44                          ^~~~~~~~~~~~~
     45                          curexc_traceback
     46     /Users/healther/Documents/git.nosync/spack/opt/spack/darwin-mojave-broadwell/clang-10.0.0-apple/python-3.7.4-sbzjkpqd5iukn357yc53xkgqkrybgfdh/include/python3.7m/pystate.h:241:15: note: 'curex
            c_traceback' declared here
     47         PyObject *curexc_traceback;
     48                   ^
  >> 49     _line_profiler.c:7902:13: error: no member named 'exc_type' in 'struct _ts'
     50         tstate->exc_type = type;
     51         ~~~~~~  ^
  >> 52     _line_profiler.c:7903:13: error: no member named 'exc_value' in 'struct _ts'; did you mean 'curexc_value'?
     53         tstate->exc_value = value;
     54                 ^~~~~~~~~
     55                 curexc_value
     56     /Users/healther/Documents/git.nosync/spack/opt/spack/darwin-mojave-broadwell/clang-10.0.0-apple/python-3.7.4-sbzjkpqd5iukn357yc53xkgqkrybgfdh/include/python3.7m/pystate.h:240:15: note: 'curex
            c_value' declared here
     57         PyObject *curexc_value;
     58                   ^
  >> 59     _line_profiler.c:7904:13: error: no member named 'exc_traceback' in 'struct _ts'; did you mean 'curexc_traceback'?
     60         tstate->exc_traceback = tb;
     61                 ^~~~~~~~~~~~~
     62                 curexc_traceback
     63     /Users/healther/Documents/git.nosync/spack/opt/spack/darwin-mojave-broadwell/clang-10.0.0-apple/python-3.7.4-sbzjkpqd5iukn357yc53xkgqkrybgfdh/include/python3.7m/pystate.h:241:15: note: 'curex
            c_traceback' declared here
     64         PyObject *curexc_traceback;
     65                   ^
  >> 66     _line_profiler.c:7959:24: error: no member named 'exc_type' in 'struct _ts'
     67         tmp_type = tstate->exc_type;
     68                    ~~~~~~  ^
  >> 69     _line_profiler.c:7960:25: error: no member named 'exc_value' in 'struct _ts'; did you mean 'curexc_value'?
     70         tmp_value = tstate->exc_value;
     71                             ^~~~~~~~~
     72                             curexc_value
     73     /Users/healther/Documents/git.nosync/spack/opt/spack/darwin-mojave-broadwell/clang-10.0.0-apple/python-3.7.4-sbzjkpqd5iukn357yc53xkgqkrybgfdh/include/python3.7m/pystate.h:240:15: note: 'curex
            c_value' declared here
     74         PyObject *curexc_value;
     75                   ^
  >> 76     _line_profiler.c:7961:22: error: no member named 'exc_traceback' in 'struct _ts'; did you mean 'curexc_traceback'?
     77         tmp_tb = tstate->exc_traceback;
     78                          ^~~~~~~~~~~~~
     79                          curexc_traceback
     80     /Users/healther/Documents/git.nosync/spack/opt/spack/darwin-mojave-broadwell/clang-10.0.0-apple/python-3.7.4-sbzjkpqd5iukn357yc53xkgqkrybgfdh/include/python3.7m/pystate.h:241:15: note: 'curex
            c_traceback' declared here
     81         PyObject *curexc_traceback;
     82                   ^
  >> 83     _line_profiler.c:7962:13: error: no member named 'exc_type' in 'struct _ts'
     84         tstate->exc_type = local_type;
     85         ~~~~~~  ^
  >> 86     _line_profiler.c:7963:13: error: no member named 'exc_value' in 'struct _ts'; did you mean 'curexc_value'?
     87         tstate->exc_value = local_value;
     88                 ^~~~~~~~~
     89                 curexc_value
     90     /Users/healther/Documents/git.nosync/spack/opt/spack/darwin-mojave-broadwell/clang-10.0.0-apple/python-3.7.4-sbzjkpqd5iukn357yc53xkgqkrybgfdh/include/python3.7m/pystate.h:240:15: note: 'curex
            c_value' declared here
     91         PyObject *curexc_value;
     92                   ^
  >> 93     _line_profiler.c:7964:13: error: no member named 'exc_traceback' in 'struct _ts'; did you mean 'curexc_traceback'?
     94         tstate->exc_traceback = local_tb;
     95                 ^~~~~~~~~~~~~
     96                 curexc_traceback
     97     /Users/healther/Documents/git.nosync/spack/opt/spack/darwin-mojave-broadwell/clang-10.0.0-apple/python-3.7.4-sbzjkpqd5iukn357yc53xkgqkrybgfdh/include/python3.7m/pystate.h:241:15: note: 'curex
            c_traceback' declared here
     98         PyObject *curexc_traceback;
     99                   ^
     100    15 errors generated.
  >> 101    error: command '/Users/healther/Documents/git.nosync/spack/lib/spack/env/clang/clang' failed with exit status 1
```

I added the extra version, but it doesn't alleviate the problem of not having released support for the current (or next-to-current python version)

Reply:
This seems to be a known issue. See https://github.com/rkern/line_profiler/issues/127 and https://github.com/rkern/line_profiler/issues/132 (and like dozens of other issues on their GitHub...). Can you see if [this hack](https://github.com/rkern/line_profiler/issues/132#issuecomment-418913772) works on the latest release?

Reply:
I didn't need to do the cleaning in the first place with just using the commit above (which currently corresponds to the HEAD, but allows for prefetching and is a stable result).

Reply:
Yes, but if you apply the same hack to the 2.1.2 release, we won't have to make our own release from some random commit. Based on https://github.com/pandas-dev/pandas/issues/21785 and https://github.com/numpy/numpy/issues/10500 it seems to be a common issue when using an older version of cython with Python 3.7. You may need to update the version of cython in Spack.

Reply:
FWIW, this fixes the compilation with Python 3.7, following the README:
https://github.com/BlueBrain/spack/blob/496c783c16a559181dcffee1f9274a1747070b03/var/spack/repos/builtin/packages/py-line-profiler/package.py#L25-L33
Wouldn't this be better than dropping Python 3 support?

Reply:
I don't mind if we merge @matz-e solution either. I really just want a working line profiler^^ care to make a PR out of that? 

That being said: I'm not sure how bad a "random commit" is for spack. It won't be preferred either way and there are multiple commits since the last line-profiler release. As long as we don't depend on the branch (which can change under us) but use a fixed commit we don't break anything in spack's model?

Reply:
> care to make a PR out of that?

See #13102… it's building in our deployment for now (with Python 3.7).